cmd/helm/completion.go

```diff
@@ -74,173 +74,23 @@ func runCompletion(out io.Writer, cmd *cobra.Command, args []string) error {
 	if !found {
 		return errors.Errorf("unsupported shell type %q", args[0])
 	}
+	// In case the user renamed the helm binary (e.g., to be able to run
+	// both helm2 and helm3), we overwrite the new binary name to the command
+	// in order to generate proper completion
+	rootCmd := cmd.Root()
+	rootCmd.Use = filepath.Base(os.Args[0])
 
-	return run(out, cmd)
+	return run(out, rootCmd)
 }
 
 func runCompletionBash(out io.Writer, cmd *cobra.Command) error {
-	err := cmd.Root().GenBashCompletion(out)
-
-	// In case the user renamed the helm binary (e.g., to be able to run
-	// both helm2 and helm3), we hook the new binary name to the completion function
-	if binary := filepath.Base(os.Args[0]); binary != "helm" {
-		renamedBinaryHook := `
-# Hook the command used to generate the completion script
-# to the helm completion function to handle the case where
-# the user renamed the helm binary
-if [[ $(type -t compopt) = "builtin" ]]; then
-    complete -o default -F __start_helm %[1]s
-else
-    complete -o default -o nospace -F __start_helm %[1]s
-fi
-`
-		fmt.Fprintf(out, renamedBinaryHook, binary)
-	}
-
-	return err
+	return cmd.GenBashCompletion(out)
 }
 
 func runCompletionZsh(out io.Writer, cmd *cobra.Command) error {
-	zshInitialization := `#compdef helm
-
-__helm_bash_source() {
-	alias shopt=':'
-	alias _expand=_bash_expand
-	alias _complete=_bash_comp
-	emulate -L sh
-	setopt kshglob noshglob braceexpand
-	source "$@"
-}
-__helm_type() {
-	# -t is not supported by zsh
-	if [ "$1" == "-t" ]; then
-		shift
-		# fake Bash 4 to disable "complete -o nospace". Instead
-		# "compopt +-o nospace" is used in the code to toggle trailing
-		# spaces. We don't support that, but leave trailing spaces on
-		# all the time
-		if [ "$1" = "__helm_compopt" ]; then
-			echo builtin
-			return 0
-		fi
-	fi
-	type "$@"
-}
-__helm_compgen() {
-	local completions w
-	completions=( $(compgen "$@") ) || return $?
-	# filter by given word as prefix
-	while [[ "$1" = -* && "$1" != -- ]]; do
-		shift
-		shift
-	done
-	if [[ "$1" == -- ]]; then
-		shift
-	fi
-	for w in "${completions[@]}"; do
-		if [[ "${w}" = "$1"* ]]; then
-			# Use printf instead of echo beause it is possible that
-			# the value to print is -n, which would be interpreted
-			# as a flag to echo
-			printf "%s\n" "${w}"
-		fi
-	done
-}
-__helm_compopt() {
-	true # don't do anything. Not supported by bashcompinit in zsh
+	return cmd.GenZshCompletion(out)
```

Comment:
Cobra 1.0's support for Zsh is a lesser solution than what helm currently provides.
The biggest issue is that it does not support custom completions, so
`helm status <TAB>` would no longer work, for example.

We first need https://github.com/spf13/cobra/pull/1070 to be merged and released before we can start using Cobra's Zsh completion.

Reply:
Agreed